pkg/kv/kvserver/batcheval/cmd_export.go

```diff
@@ -249,21 +277,39 @@ func evalExport(
 			// `Fingerprint` for point-keys and the SST file that contains the
 			// rangekeys we encountered during ExportRequest evaluation.
 			exported = roachpb.ExportResponse_File{
-				EndKeyTS:    resume.Timestamp,
+				EndKeyTS:    resumeInfo.ResumeKey.Timestamp,
 				SST:         data,
 				Fingerprint: fingerprint,
 			}
 		} else {
 			exported = roachpb.ExportResponse_File{
 				Span:     span,
-				EndKeyTS: resume.Timestamp,
+				EndKeyTS: resumeInfo.ResumeKey.Timestamp,
 				Exported: summary,
 				SST:      data,
 			}
 		}
 		reply.Files = append(reply.Files, exported)
-		start = resume.Key
-		resumeKeyTS = resume.Timestamp
+		start = resumeInfo.ResumeKey.Key
+		resumeKeyTS = resumeInfo.ResumeKey.Timestamp
+
+		// If we paginated because we are over our allotted CPU limit, we must break
+		// from command evaluation and return a response to the client before
+		// resuming our export from the resume key. This gives the scheduler a
+		// chance to take the current goroutine off CPU and allow other processes to
+		// progress.
+		if resumeInfo.CPUOverlimit {
+			if resumeInfo.ResumeKey.Key != nil {
+				reply.ResumeSpan = &roachpb.Span{
+					Key:    resumeInfo.ResumeKey.Key,
+					EndKey: args.EndKey,
+				}
+				// TODO(during review): Do we want to add another resume reason
+				// specifically for CPU preemption.
+				reply.ResumeReason = roachpb.RESUME_UNKNOWN
```

Comment:
I didn't spot anywhere that a new reasons would cause a problem. Very few locations seem to even read the reason.  My vote would be to add a reason since we do know the reason here.

Reply:
done!